File: model/contract_deployer.py
import sys
from pathlib import Path

project_root = Path(__file__).parent.parent.parent
sys.path.append(str(project_root))



import algokit_utils
from algokit_utils.beta.algorand_client import AlgorandClient, PayParams

from smart_contracts.artifacts.moderator.moderator_client import (
    ModeratorClient,
)


def deploy(
    ipfsHash: str,
    modelParameters: str,
) -> None:
    algorand = AlgorandClient.default_local_net()
    algorand.set_default_validity_window(1000)

    dispenser = algorand.account.dispenser()

    acct = algorand.account.random()

    algorand.send.payment(
        PayParams(sender=dispenser.address, receiver=acct.address, amount=10_000_000),
    )

    app_client = ModeratorClient(
        algod_client=algorand.client.algod,
        creator=acct.address,
        indexer_client=algorand.client.indexer,
        signer=acct.signer,
    )

    print(f"deploying: {app_client.app_id}")

    app_client.deploy(
        on_schema_break=algokit_utils.OnSchemaBreak.AppendApp,
        on_update=algokit_utils.OnUpdate.AppendApp,
    )

    response = app_client.hash(ipfsHash=ipfsHash)
    params = app_client.model_parameters(modelParameters=modelParameters)
    print(f"{app_client.app_id} received: {response.return_value}")
    print(f"{app_client.app_id} received: {params.return_value}")

File: projects/DML-contracts/contracts/ContractDeployer.ts
import { AlgorandClient } from '@algorandfoundation/algokit-utils';
import * as algokit from '@algorandfoundation/algokit-utils';
import { encodeAddress } from 'algosdk';
import { DmlChainFactory } from './clients/DMLChainClient';

type Classification = {
  accuracy: bigint;
  precision: bigint;
  recall: bigint;
  f1score: bigint;
};

type ParamsData = {
  paramHash: string;
  paramKey: string;
};

export const contractDeployer = async (
  ipfsHash: string,
  modelEvaluation: Classification,
  ParameterData: ParamsData
) => {
  console.log('these are params', ParameterData);
  const algorand = AlgorandClient.defaultLocalNet();
  algorand.setDefaultValidityWindow(1000);

  const dispenser = await algorand.account.localNetDispenser();

  algokit.Config.configure({
    populateAppCallResources: true,
  });

  const acct = algorand.account.random();

  algorand.send.payment({
    sender: dispenser.addr,
    receiver: acct.account.addr,
    amount: (10).algo(),
  });

  const factory = algorand.client.getTypedAppFactory(DmlChainFactory, {
    defaultSender: acct.account.addr,
  });

  algokit.Config.configure({
    populateAppCallResources: true,
  });

  const { appClient: client } = await factory.send.create.createApplication({ args: { modelHash: ipfsHash } });

  const appID = client.appId;

  console.log('this is app ID', appID);

  const mbrPayFirstDeposit = await algorand.createTransaction.payment({
    sender: acct.account.addr,
    receiver: client.appAddress,
    amount: (1).algo(),
  });

  const createBox = await client.send
    .storeClassificationSelectionCriteria({
      args: {
        evaluationMetrics: modelEvaluation,
        mbrPay: mbrPayFirstDeposit,
      },
    })
    .then((response) => response.confirmation);

  console.log('this is create box', createBox);

  const getBox = await client.send.getClassificationCriteria();

  console.log('this is getbox', getBox.return);

  const performMetricEvaluation = await client.send.classModelSelectionCriteria({
    args: { modelEvaluationMetrics: modelEvaluation },
  });

  console.log(performMetricEvaluation.return);

  const modelParamBox = await algorand.createTransaction.payment({
    sender: acct.account.addr,
    receiver: client.appAddress,
    amount: (1).algo(),
  });

  const storeModelParameters = await client.send.storeModelParams({
    args: { mbrPay: modelParamBox, address: acct.account.addr, paramsData: ParameterData },
  });

  console.log('storing model params', storeModelParameters);

  const boxIDs = await algorand.app.getBoxNames(appID);

  const boxes = await client.appClient.getBoxNames();
  console.log('Boxes:', boxes);

  console.log('ÍDs', boxIDs);

  boxIDs.forEach(async (box) => {
    if (Object.keys(box.nameRaw).length === 32) {
      try {
        const extAddr = encodeAddress(box.nameRaw);
        console.log(extAddr, 'this is ext addr');
        const getParams = await (await client.send.getBoxValue({ args: { address: extAddr } })).return;
        const paramHash = getParams?.paramHash;
        const paramKey = getParams?.paramKey;
        console.log(paramHash, paramKey);
      } catch (error) {
        console.error(`Error fetching box value for ${box.name}`, error);
      }
    } else {
      console.warn(`Skipped API call for box with name: ${box.name} due to being <32 for address type`);
    }
  });
};
